### PR TITLE
Add role for setup-sources

### PR DESCRIPTION
**Pull request by pcreech**
Add role for setup-sources

Fixes GH-3

### CODE REVIEW

File: obal/data/roles/setup-sources/tasks/defaults/main.yaml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+---
+force: false
+annex_opts: ""
```
Comment (ehelms) [inline obal/data/roles/setup-sources/tasks/defaults/main.yaml]: Convention says to prefix variables with the role to namespace them and prevent clashes.

File: obal/data/roles/setup-sources/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,77 @@
+---
+- set_fact:
+    pkg_dir: "{{ inventory_dir }}/packages/{{ inventory_hostname }}"
+    annex_opts: annex_opts
```
Comment (ehelms) [inline obal/data/roles/setup-sources/tasks/main.yml]: I can live with the shortening of words if others are, but generally I tend towards explicit verbosity (e.g. package over pkg and options over opts).

Comment (ehelms) [reply obal/data/roles/setup-sources/tasks/main.yml]: Does this need a tag always?

File: obal/data/roles/setup-sources/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  register: spec_file
+  tags:
+    - always
+
+- set_fact:
+    spec_file_path: "{{ spec_file.files[0].path }}"
+  tags:
+    - always
+
+- name: 'Extract Source0 from spec'
```
Comment (ehelms) [inline obal/data/roles/setup-sources/tasks/main.yml]: What happens if I have more than one source that is annexed? Rare, but possible.

File: obal/data/roles/setup-sources/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    source0: "{{ source0.stdout }}"
+  tags:
+    - always
+
+- name: 'Extract sourcebase from Source0'
+  shell: "basename {{ source0 }}"
+  register: sourcebase
+  changed_when: False
+  tags:
+    - always
```
Comment (ehelms) [inline obal/data/roles/setup-sources/tasks/main.yml]: I think you can combine this with the `set_fact` above using the `basename` filter Ansible provides

File: obal/data/roles/setup-sources/tasks/main.yml (hunk 3 above)
Comment (pcreech) [reply obal/data/roles/setup-sources/tasks/main.yml]: In this case, it will be ignored.  Based on my reading of the setup_sources.sh file I was basing this off of, this was the only case.  I can go back and add in support for multiple sources

Comment (ekohl) [reply obal/data/roles/setup-sources/tasks/main.yml]: https://github.com/theforeman/foreman-packaging/blob/rpm/develop/setup_sources.sh does loop over the files

File: obal/data/roles/setup-sources/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,59 @@
+---
+- set_fact:
+    package_dir: "{{ inventory_dir }}/packages/{{ inventory_hostname }}"
+    setup_sources_annex_options: setup_sources_annex_options
```
Comment (ekohl) [inline obal/data/roles/setup-sources/tasks/main.yml]: Why is this needed?

File: obal/data/roles/setup-sources/tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  register: search_annex
+  ignore_errors: yes
+  when: setup_sources_force or ( source_exists.stat.islink is defined and source_exists.stat.islnk )
+
+- name: 'Add url to annex for source'
+  shell: 'git annex addurl {{ setup_sources_annex_options }} --file "{{ sourcebase }}" "{{ source0 }}"'
+  args:
+    chdir: "{{ package_dir }}"
+  ignore_errors: yes
+  when: search_annex.rc is defined and search_annex.rc != 0
```
Comment (ekohl) [inline obal/data/roles/setup-sources/tasks/main.yml]: I'm still quite new to ansible but would ```search_annex|failed``` work?

File: obal/data/roles/setup-sources/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+  tags:
+    - always
+
+- name: 'Source exists'
+  stat: 
+    path: "{{ sourcepath }}"
+  register: source_exists
+
+- name: 'Search annex for web link'
+  shell: 'git annex whereis {{ sourcepath }} 2>/dev/null | grep -q " web"'
```
Comment (ekohl) [inline obal/data/roles/setup-sources/tasks/main.yml]: foreman-packaging is a bit more specific in that it tests for ``` web:```

**Comment by ekohl** [comment]
Should ```tasks/setup_sources.yml``` be removed too?

File: obal/data/roles/setup-sources/tasks/main.yml (hunk 5 above)
Comment (evgeni) [reply obal/data/roles/setup-sources/tasks/main.yml]: it's not, esp it sets the `setup_sources_annex_options` fact to the literal `setup_sources_annex_options` string

File: obal/data/roles/setup-sources/tasks/main.yml (hunk 3 above)
Comment (evgeni) [reply obal/data/roles/setup-sources/tasks/main.yml]: "our" version was ignoring other sources so far. how does the foreman code cope with non-remote files, or is `[ -h ... ]` enough to catch those?

**Comment by evgeni** [comment]
When I added `-f` (aka `setup_sources_force`) to *our* copy of `setup_sources.sh`, the idea was that I want the script to do `git annex addurl` also for *new* files (so `[ -h ...]` was false).

now, a year later, I think the check also could have been:
```bash
  [ -h $d/$sourcebase ] || [ ! -e $d/$sourcebase ] || continue
```

is it a link? cool! does it not exist? cool. otherwise, don't touch this.

`git annex whereis` fails on a non-existing path, though, so we'd have to catch that.

File: obal/data/roles/setup-sources/tasks/main.yml (hunk 3 above)
Comment (ekohl) [reply obal/data/roles/setup-sources/tasks/main.yml]: I haven't seen issues with the actually included tarballs we have with bundled installations. According to ```man test```:
> -h FILE
> FILE exists and is a symbolic link (same as -L)

That implies it's never an issue with actual files.

File: obal/data/roles/setup-sources/tasks/main.yml (hunk 5 above)
Comment (pcreech) [reply obal/data/roles/setup-sources/tasks/main.yml]: Yeah, I think this snuck in there b/c I had the defaults in the wrong place and I didn't remove it

File: obal/data/roles/setup-sources/tasks/main.yml (hunk 6 above)
Comment (ehelms) [reply obal/data/roles/setup-sources/tasks/main.yml]: Yup.

**Comment by ehelms** [comment]
General question given this is the first new role being contributed. I believe convention is to do roles with underscores in their names and thus their folder structure. Should we switch to that?

**Comment by evgeni** [comment]
:+1: for python naming

**Comment by pcreech** [comment]
Since it was simple to convert the spectool output into iterable items, I was able to refactor the role to support multiple source lines.

**Comment by ehelms** [comment]
I am a bit confused on why the role moved under `obal/data/roles` ?

**Comment by pcreech** [comment]
Re: move under obal/data/roles... looks like it was done as part of https://github.com/evgeni/obal/commit/b8469f51ab6da2dec99cead83a3d46d539ec1cf3

**Comment by evgeni** [comment]
Yeah. That was the easiest way to distribute the roles as part of thr python  package. We can rethink that if you want to.

**Comment by evgeni** [comment]
Yeah. That was the easiest way to distribute the roles as part of thr python  package. We can rethink that if you want to.

**Comment by evgeni** [comment]
Yeah. That was the easiest way to distribute the roles as part of thr python  package. We can rethink that if you want to.

**Comment by ehelms** [comment]
@evgeni I think it just surprised me as a change, and my first impression was how it hid them under a deep path that was not that discoverable or obvious - still noodling on it.

@ekohl Final review falls to you

**Comment by ekohl** [comment]
> @evgeni I think it just surprised me as a change, and my first impression was how it hid them under a deep path that was not that discoverable or obvious - still noodling on it.

We could do the same as in foreman-infra and add a symlink at the root level.

File: obal/data/roles/setup_sources/tasks/annex.yml [hunk 8]
```diff
@@ -0,0 +1,32 @@
+---
+- set_fact:
+    source0: "{{ item }}"
```
Comment (ekohl) [inline obal/data/roles/setup_sources/tasks/annex.yml]: Just source might be cleaner?
